File: js/problem18.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */

let res = null;

function twoSum(nums, target, currentSum, start) {
    let i = start + 1, j = nums.length - 1;
    while (i < j) {
        let temp = currentSum + nums[start] + nums[i] + nums[j];
        if (temp === target) {
            res.push([currentSum, nums[start], nums[i], nums[j]]);
            i++;
            while ((i < j) && (nums[i] === nums[i - 1])) {
                i++;
            }
            j--;
            while ((i < j) && (nums[j] === nums[j + 1])) {
                j--;
            }
        } else if (temp < target) {
            i++;
        } else {
            j--;
        }
    }
}

function threeSum(nums, target, start) {
    for (let i = start + 1; i < nums.length - 2; i++) {
        if ((i > start + 1) && (nums[i] === nums[i - 1])) {
            continue;
        }
        twoSum(nums, target, nums[start], i);
    }
}

var fourSum = function(nums, target) {
    if (nums.length < 4) {
        return [];
    }
    
    res = [];
    
    nums.sort(function(a, b) {
        return (+a) - (+b); 
    });
    
    for (let i = 0; i < nums.length - 3; i++) {
        if ((i > 0) && (nums[i] === nums[i - 1])) {
            continue;
        }
        threeSum(nums, target, i);
    }
    
    return res;
    
};

File: leetcodeJava/src/main/java/Problem18.java
// 18. 4Sum

class Solution {
    
    private void twoSum(int[] nums, int current, int previousSum, List<List<Integer>> res, int target) {
        int i = current + 1, j = nums.length - 1;
        while (i < j) {
            int tempSum = previousSum + nums[current] + nums[i] + nums[j];
            if (tempSum == target) {
                List<Integer> tempList = new ArrayList<Integer>();
                tempList.add(previousSum);
                tempList.add(nums[current]);
                tempList.add(nums[i]);
                tempList.add(nums[j]);
                res.add(tempList);
                i++;
                while ((i < j) && (nums[i] == nums[i - 1])) {
                    i++;
                }
                j--;
                while ((i < j) && (nums[j] == nums[j + 1])) {
                    j--;
                }
            } else if (tempSum < target) {
                i++;
            } else {
                j--;
            }
        }
    }
    
    private void threeSum(int[] nums, int current, List<List<Integer>> res, int target) {
        for (int i = current + 1; i < nums.length - 2; i++) {
            if ((i > current + 1) && (nums[i] == nums[i - 1])) {
                continue;
            }
            
            this.twoSum(nums, i, nums[current], res, target);
        }
    }
    
    public List<List<Integer>> fourSum(int[] nums, int target) {
        if (nums.length < 3) {
            return new ArrayList<List<Integer>>();
        }
        
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 3; i++) {
            if ((i > 0) && (nums[i] == nums[i - 1])) {
                continue;
            }
            
            this.threeSum(nums, i, res, target);
        }
        
        return res;
    }
}